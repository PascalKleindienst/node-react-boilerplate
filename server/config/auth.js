// =========================================================================
// Configuration for different Auth Providers===============================
// =========================================================================
module.exports = {
    // https://apps.twitter.com/
    'twitterAuth' : {
        'consumerKey'    : process.env.TWITTER_CONSUMER_KEY,
        'consumerSecret' : process.env.TWITTER_CONSUMER_SECRET,
        'callbackURL'    : process.env.TWITTER_CALLBACK_URL,
        'userProfileURL' : 'https://api.twitter.com/1.1/account/verify_credentials.json?include_email=true',
    },

    // https://console.developers.google.com/apis/credentials
    'googleAuth' : {
        'clientID'     : process.env.GOOGLE_CLIENT_ID,
        'clientSecret' : process.env.GOOGLE_CLIENT_SECRET,
        'callbackURL'  : process.env.GOOGLE_CALLBACK_URL
    },

    // https://developers.facebook.com/
    'facebookAuth' : {
        'clientID'      : process.env.FACEBOOK_CLIENT_ID,
        'clientSecret'  : process.env.FACEBOOK_CLIENT_SECRET,
        'callbackURL'   : process.env.FACEBOOK_CALLBACK_URL,
        'profileURL'    : 'https://graph.facebook.com/v2.5/me?fields=first_name,last_name,email',
        'profileFields' : ['id', 'email', 'name', 'picture.type(large)'] // For requesting permissions from Facebook API
    },
};
