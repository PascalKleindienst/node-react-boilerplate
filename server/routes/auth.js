// ==================================================
// Setup Auth Routes ================================
// ==================================================

/**
 * Setup Auth Routes
 * @param  {Object} app
 * @param  {Object} passport
 */
module.exports = (app, passport) => {
    const redirects = {
        successRedirect : '/dashboard',
        failureRedirect : '/'
    };

    // =========================================
    // GENERAL AUTH ROUTES =====================
    // =========================================
    app.get('/auth/logout', (req, res) => {
        req.logout();
        res.json({ message: 'Logged out user' });
    });

    app.get('/auth/user', (req, res) => {
        if (req.isAuthenticated()) {
            req.user.token = undefined;
            return res.json(req.user);
        }

        return res.json({ error: 'User is not authenticated' });
    });

    // =====================================
    // FACEBOOK ROUTES =====================
    // =====================================
    app.get('/auth/facebook', passport.authenticate('facebook', { scope : ['public_profile', 'email'] }));
    app.get('/auth/facebook/callback', passport.authenticate('facebook', redirects));

    // =====================================
    // TWITTER ROUTES ======================
    // =====================================
    app.get('/auth/twitter', passport.authenticate('twitter'));
    app.get('/auth/twitter/callback', passport.authenticate('twitter', redirects));

    // =====================================
    // GOOGLE ROUTES =======================
    // =====================================
    app.get('/auth/google', passport.authenticate('google', { scope : ['profile', 'email'] }));
    app.get('/auth/google/callback', passport.authenticate('google', redirects));
};
