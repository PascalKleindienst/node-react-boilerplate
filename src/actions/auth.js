// ==============================================
// AUTH ACTIONS =================================
// ==============================================
import ajax from 'fetchival';

// Login
export const LOGIN = 'LOGIN';
export const login = (user) => ({
    type: LOGIN,
    user
});

// Logout
export const LOGOUT = 'LOGOUT';
export const logout = () => {
    return (dispatch) => {
        return ajax('/auth/logout', { mode: 'cors', credentials: 'same-origin' }).get().then( () => {
            dispatch({
                type: LOGOUT
            });
        });
    };
};
