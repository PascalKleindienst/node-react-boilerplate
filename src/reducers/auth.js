// ==============================================
// AUTH REDUCERS ================================
// ==============================================
import { LOGIN, LOGOUT } from '../actions/auth';

export default (state = {}, action) => {
    switch (action.type) {
    case LOGIN:
        return action.user;
    case LOGOUT:
        return {};
    default:
        return state;
    }
};
