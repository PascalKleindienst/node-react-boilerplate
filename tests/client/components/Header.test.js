// ==============================================
// HEADER PAGE TEST =============================
// ==============================================
// Mocks
import ajax from 'fetchival';
jest.mock('fetchival');

// Dependencies
import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import HeaderContainer, { Header } from '../../../src/components/Header';
import { logout } from '../../../src/actions/auth';

const mockStore = configureMockStore([thunk]);

describe('my header component', () => {
    let wrapper, store;

    beforeEach(() => {
        ajax.mockImplementation(() => ({
            get: jest.fn(() => new Promise((resolve, reject) => {
                resolve({ id: 1 });
            }))
        }));

        store = mockStore({
            auth: { picture: 'foo.png' }
        });
        
        wrapper = shallow(<HeaderContainer store={store} />);
    });

    test('should render correctly', () => {
        const wrapper = shallow(<Header logout={() => {}} auth={ {picture: 'foo.png'} } />);
        expect(wrapper).toMatchSnapshot();
    });
    
    test('should call logout on button click', () => {
        const logout = jest.fn();
        const wrapper = shallow(<Header logout={logout} auth={ {picture: 'foo.png'} } />);
        
        wrapper.find('.logout').simulate('click');
        expect(logout).toHaveBeenCalled();
    });

    test('should contain auth information', () => {
        expect(wrapper.props().auth).toEqual({ picture: 'foo.png' });
    });

    test('should logout when button is clicked', () => {
        return store.dispatch(wrapper.props().logout).then(() => {
            const actions = store.getActions();
            expect(actions[0]).toEqual({ type: 'LOGOUT' });
        });
    });
});
