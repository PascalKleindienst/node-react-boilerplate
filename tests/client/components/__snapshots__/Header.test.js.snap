// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my header component should render correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/dashboard"
      >
        <h1>
          Boilerplate
        </h1>
      </Link>
      <div
        className="profile has-dropdown"
      >
        <img
          alt=""
          className="profile__picture"
          src="foo.png"
        />
        <div
          className="dropdown-menu"
        >
          <div
            className="dropdown-menu__header"
          >
            <strong>
              Account
            </strong>
          </div>
          <Link
            className="dropdown-menu__item"
            replace={false}
            to="/profile"
          >
            Profile
          </Link>
          <Link
            className="dropdown-menu__item logout"
            onClick={[Function]}
            replace={false}
            to="#"
          >
            Logout
          </Link>
        </div>
      </div>
    </div>
  </div>
</header>
`;
